core/services/workflows/metering/metering.go

```
@@ -166,67 +207,93 @@ func (r *Report) MedianSpend() map[SpendUnit]SpendValue {
+	}
+	err := r.balance.Minus(amount)
+	if err != nil {
+		return 0, err
 	}
 
-	// TODO: checking capability type to know ahead of time what to reserve is brittle
-	switch capInfo.CapabilityType {
-	case "streams-trigger@1.0.0":
-		req.Credits = []*billing.AccountCreditsInput{
-			{Credits: 1, CreditType: triggerCreditType},
-		}
-	case "cron-trigger@1.0.0":
-		req.Credits = []*billing.AccountCreditsInput{
-			{Credits: 1, CreditType: triggerCreditType},
-		}
-	case "offchain_reporting@1.0.0":
-		req.Credits = []*billing.AccountCreditsInput{
-			{Credits: 100_000, CreditType: consensusCreditType}, // TODO: reserve the maximum byte size
-		}
-	case "write_aptos-testnet@1.0.0":
-		req.Credits = []*billing.AccountCreditsInput{
-			{Credits: 100_000_000, CreditType: gasCreditType}, // TODO: reserve the maximum byte size
-		}
-	default:
-		r.lggr.Infof("unexpected capability type: %s", capInfo.CapabilityType)
+	r.steps[ref] = ReportStep{
+		Reserve: amount,
+		Spend:   nil,
 	}
 
-	// TODO: handle error by indicating that a billing reserve failed
-	// but do not halt the workflow
-	if _, err := r.client.ReserveCredits(context.TODO(), &req); err != nil {
-		r.lggr.Warnf("failed to reserve credits: %s", err)
+	return amount, nil
+}
+
+// ReserveByAvailability earmarks an amount of local universal credit balance and then returns that amount
+// The amount reserved is determined splitting the total open balance by how many remaining concurrent calls can be made
+// We expect to only set this value once - an error is returned if a step would be overwritten
+func (r *Report) ReserveByAvailability(ref string, capInfo capabilities.CapabilityInfo, openConcurrentCallSlots int) (int64, error) {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+
+	if !r.ready {
+		return 0, errors.New("metering report has not been initialized")
 	}
 
-	// TODO: handle extra reserves for write step
+	if _, ok := r.steps[ref]; ok {
+		return 0, errors.New("step reserve already exists")
```

We can return the exported error here